enhancements/ingress/add-headers-to-haproxyconfig-via-ingress-controller-or-route-api.md

```diff
@@ -0,0 +1,413 @@
+	//
+	// Note that this option only applies to cleartext HTTP connections
+	// and to secure HTTP connections for which the ingress controller
+	// terminates encryption (that is, edge-terminated or reencrypt
+	// connections).  Headers cannot be captured for TLS passthrough
+	// connections.
+	//
+	// +kubebuilder:validation:Optional
+	// +optional
+	HTTPSetHeaders []RouteSetHTTPHeaders `json:"httpSetHeaders,omitempty"`
+}
+```
+
+The `RouteSetHTTPHeaders` type has two fields `Name` and `Value` to specify header name and header value respectively.
+```go
+
+// RouteSetHTTPHeaders describes an HTTP header that should be
+// captured.
+type RouteSetHTTPHeaders struct {
+// request specifies which HTTP request headers to capture.
+//
+// If this field is empty, no request headers are captured.
+//
+// +nullable
+// +optional
+Request []RouteSetHTTPHeader `json:"request,omitempty"`
+
+// response specifies which HTTP response headers to capture.
+//
+// If this field is empty, no response headers are captured.
+//
+// +nullable
+// +optional
+Response []RouteSetHTTPHeader `json:"response,omitempty"`
+}
+```
+
+The RouteSetHTTPHeader type has fields for specifying a header name, header value and action.
+```go
+type RouteSetHTTPHeader{
+	// name specifies a header name.  Its value must be a valid HTTP header
+	// name as defined in RFC 2616 section 4.2.
+	//
+	// +kubebuilder:validation:Required
+    // +kubebuilder:validation:MinLength=1
+    // +kubebuilder:validation:MaxLength=256
+    // +kubebuilder:validation:Pattern=`^[A-Za-z0-9!#$%&'*+\-.^_\x60|~]+$`
+	// +kubebuilder:validation:Pattern="^[-!#$%&'*+.0-9A-Z^_`a-z|~]+$"
+	// +required
+	Name string `json:"name"`
+
+	// Value specifies a header value.
+	// +kubebuilder:validation:Required
+	// +required
+    // +kubebuilder:validation:MinLength=1
+    // +kubebuilder:validation:MaxLength=4096
+	Value string `json:"value"`
+
+    // Action specifies a header value.
+    // +kubebuilder:validation:Required
+    // +required
+    // +kubebuilder:validation:Enum=Append;Replace
+    Action IngressControllerHTTPHeaderPolicy `json:"action"`
+}
+```
+
+
+### Implementation Details/Notes/Constraints
+A) Setting HTTP headers by cluster admin via IngressController's Spec.
+
+Implementing this enhancement requires changes in the following repositories:
+
+* openshift/api
+* openshift/cluster-ingress-operator
+* openshift/router
+
+The router configures HAProxy using a configuration template.  The template uses
+environment variables as input parameters.  The enhancement adds a new
+environment variables, `ROUTER_SET_HTTP_HEADERS`, which specify the headers to set in the
+frontend fe_no_sni, frontend public, frontend fe_sni  section of the haproxy.config.
+The ingress operator sets this variable based on the IngressController's
+`spec.httpSetHeaders` field value.
+
+The values of `ROUTER_SET_HTTP_HEADERS` are comma-delimited lists of base64 encoded values of
+the form `name:value` separated by comma.
```

Comment:
In https://github.com/openshift/enhancements/pull/1296#discussion_r1047688239, I suggested using percent encoding (as used in URIs and [defined in RFC 3986](https://datatracker.ietf.org/doc/html/rfc3986#section-2.1)), which has the advantages that it is familiar to anyone who works with HTTP and that usually most of the characters in a percent-encoded string don't get changed and remain human-readable.  Is there a compelling reason to use base64 encoding instead?

Reply:
with base64 encoding it will be better to delimit by comma as base64 encoded value does not have comma.

Reply:
BTW it looks like percent encoding encodes comma, so we can use comma as a delimiter

Reply:
Percent encoding can encode any character you want to encode.  * grin *.